core/domain/exp_domain.py

```diff
@@ -2213,6 +2244,224 @@ def _convert_states_v28_dict_to_v29_dict(cls, states_dict):
             state_dict['solicit_answer_details'] = False
         return states_dict
 
+    @classmethod
+    def _convert_states_v29_dict_to_v30_dict(cls, states_dict):
+        """Converts from version 29 to 30. Version 30 added image assets field
+        in state model. This funtion is extracting HTMl of state and do 3
+        operations on each HTML content. The 3 operations and their sequence
+        were.
+        1) Gets image src of all images present in a state.
+        2) Assign each image source an image id or say maps each image id with
+           the image source in image info dict.
+        3) Give each image tag an image id with the help of image source, and
+           removes filepath from it.
+        After the above 3 operations each image info present in image dict is
+        added to image assets.
+
+        Args:
+            states_dict: dict. A dict where each key-value pair represents,
+                respectively, a state name and a dict used to initialize a
+                State domain object.
+
+        Returns:
+            dict. The converted states_dict.
+        """
+        image_counter = 0
+        for state_dict in states_dict.itervalues():
+            image_assets = {}
+            image_mapping = {}
+            image_info_dict_1 = {}
+            image_info_dict_2 = {}
+            image_src_list = []
+            # After extracting image source of each image, we assign image id
+            # to image tag present in html in image_info_dict_1. And we find
+            # image id of image tag with the help of image source. But there is
+            # chance that we have same image source for more then one image id,
+            # which can leads to return same and one image id for same source.
+            # So to overcome this, after assigning image id to image tag, we
+            # remove image id from image_info_dict_1 and add it to
+            # image_info_dict_2 so that we can add image info in image assets
+            # and image_info_dict_1 doesn't return same and one image_id for
+            # same image src.
+
+
+            # Gets image source, maps image source with image id in image info
+            # dict and give each image id to each image present in content html.
+            content_html = state_dict['content']['html']
+            image_src_list = image_src_list + (
+                html_validation_service.get_image_src_from_html(
```

Comment:
@import-keshav, Have you looked into this function `state_domain.State.convert_html_fields_in_state`?

Also, see exp_domain.Exploration._convert_states_v21_dict_to_v22_dict method.

Reply:
done